### PR TITLE
[ADI] Add POST/DELETE/PUT to /threads and /thread_comments

### PR DESCRIPTION
**Pull request by JeremieSamson**

### CODE REVIEW

File: src/Controller/Api/ThreadCommentController.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+
+class ThreadCommentController extends Controller
+{
+    /**
+     * @ParamConverter("threadComment", class="AppBundle\Entity\IdeasWorkshop\ThreadComment")
+     */
+    public function approveAction(ThreadComment $threadComment, ObjectManager $manager): ThreadComment
+    {
+        $status = ThreadCommentStatusEnum::APPROVED;
```
Comment (thomasbisignani) [inline src/Controller/Api/ThreadCommentController.php]: This variable is useless.

You can directly write : 

```php
$threadComment->setStatus(ThreadCommentStatusEnum::APPROVED);
```

File: src/Controller/Api/ThreadCommentController.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+        return $threadComment;
+    }
+
+    /**
+     * @ParamConverter("threadComment", class="AppBundle\Entity\IdeasWorkshop\ThreadComment")
+     */
+    public function reportAction(ThreadComment $threadComment, ObjectManager $manager): ThreadComment
+    {
+        $status = ThreadCommentStatusEnum::REPORTED;
```
Comment (thomasbisignani) [inline src/Controller/Api/ThreadCommentController.php]: Same here, useless variable.

File: src/Controller/Api/ThreadController.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+
+class ThreadController extends Controller
+{
+    /**
+     * @ParamConverter("thread", class="AppBundle\Entity\IdeasWorkshop\Thread")
+     */
+    public function approveAction(Thread $thread, ObjectManager $manager): Thread
+    {
+        $status = ThreadCommentStatusEnum::APPROVED;
```
Comment (thomasbisignani) [inline src/Controller/Api/ThreadController.php]: Same here, useless variable.

File: src/Controller/Api/ThreadController.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+        return $thread;
+    }
+
+    /**
+     * @ParamConverter("thread", class="AppBundle\Entity\IdeasWorkshop\Thread")
+     */
+    public function reportAction(Thread $thread, ObjectManager $manager): Thread
+    {
+        $status = ThreadCommentStatusEnum::REPORTED;
```
Comment (thomasbisignani) [inline src/Controller/Api/ThreadController.php]: Same here, useless variable.

File: src/Controller/Api/ThreadCommentController.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+class ThreadCommentController extends Controller
+{
+    /**
+     * @ParamConverter("threadComment", class="AppBundle\Entity\IdeasWorkshop\ThreadComment")
+     */
+    public function approveAction(ThreadComment $threadComment, ObjectManager $manager): ThreadComment
+    {
+        $status = ThreadCommentStatusEnum::APPROVED;
+
+        $threadComment->setStatus($status);
```
[the remark this answers is not in the document]
Comment (thomasbisignani) [reply src/Controller/Api/ThreadCommentController.php]: Nice idea, or directly `$entity->approve()` ?

File: src/Entity/IdeasWorkshop/Thread.php [hunk 6]
```diff
@@ -26,13 +29,38 @@
...
+ *         }
+ *     },
+ *     itemOperations={
+ *         "get",
+ *         "put_status_approve": {
+ *             "method": "PUT",
+ *             "path": "/threads/{id}/approve",
+ *             "requirements": {"id": "\d+"},
+ *             "access_control": "object.getIdeaAuthor() == user",
+ *             "denormalizationContext": {"groups": {"thread_read"}},
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/Thread.php]: why do we need `denormalizationContext` in put action?

File: src/Controller/Api/ThreadCommentController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Controller/Api/ThreadCommentController.php]: Added `approve()` and `report()` in `BaseComment` in last commit, thanks

File: src/Entity/IdeasWorkshop/Thread.php [hunk 7]
```diff
@@ -52,19 +82,29 @@ class Thread extends BaseComment implements AuthorInterface, ReportableInterface
      */
     private $comments;
 
-    public function __construct(
+    public function __construct(UuidInterface $uuid = null)
+    {
+        $this->uuid = $uuid ?: Uuid::uuid4();
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/Thread.php]: `$uuid ?? Uuid::uuid4()`

File: src/Entity/IdeasWorkshop/ThreadComment.php [hunk 8]
```diff
@@ -21,13 +24,38 @@
...
+ *             "access_control": "is_granted('ROLE_ADHERENT')",
+ *         }
+ *     },
+ *     itemOperations={
+ *         "get",
+ *         "put_status_approve": {
+ *             "method": "PUT",
+ *             "path": "/thread_comments/{id}/approve",
+ *             "requirements": {"id": "\d+"},
+ *             "denormalizationContext": {"groups": {"thread_comment_read"}},
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/ThreadComment.php]: same question for denormalization context

File: src/Entity/IdeasWorkshop/ThreadCommentStatusEnum.php [hunk 9]
```diff
@@ -15,4 +15,9 @@
         self::APPROVED,
         self::POSTED,
     ];
+
+    public const EDITABLE_STATUSES = [
```
Comment (ottaviano) [inline src/Entity/IdeasWorkshop/ThreadCommentStatusEnum.php]: do you need to use it?

Comment (JeremieSamson) [reply src/Entity/IdeasWorkshop/ThreadCommentStatusEnum.php]: nop, not needed anymore

File: src/Entity/IdeasWorkshop/BaseComment.php [hunk 10]
```diff
@@ -76,4 +63,14 @@ public static function getVisibleStatuses(): array
     {
         return ThreadCommentStatusEnum::VISIBLE_STATUSES;
     }
+
+    public function approve(): void
```
Comment (thomasbisignani) [inline src/Entity/IdeasWorkshop/BaseComment.php]: What about adding these two methods in the `EntityThreadCommentStatusTrait` ? and set directly the property like : 

```php 
public function approve(): void
{
    $this->status = ThreadCommentStatusEnum::APPROVED;
}

**Comment by thomasbisignani** [comment]
:1st_place_medal: :spaghetti:

**Pull request merged**